### PR TITLE
Feat/cuda graph training

### PR DESCRIPTION
**Pull request by EHxuban11**
## Code provenance

Original implementation written for this PR. CUDA graph capture is built entirely on PyTorch public API (torch.cuda.make_graphed_callables, torch.cuda.is_current_stream_capturing) following PyTorch documented capture recipes. No third-party training framework code was consulted or referenced.

<details><summary><h3>Greptile Summary</h3></summary>

Implements opt-in, single-GPU CUDA graph training while retaining eager loss computation and fallback behavior.
- Adds graph capture, shape-based dispatch, buffer restoration, and eager fallback infrastructure.
- Integrates capture specifications for YOLO9 and RF-DETR.
- Exposes the feature through training configuration and the CLI.
- Adds CUDA parity, routing, fallback, and AMP tests plus usage documentation.
</details>

<details><summary><h3>Confidence Score: 5/5</h3></summary>

The PR appears safe to merge.

No blocking failure remains; the previously reported BatchNorm warm-up drift is addressed by snapshotting buffers before capture, restoring them in place before the first replay, and testing exact YOLO9 buffer parity.
</details>

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| libreyolo/training/cuda_graph.py | Adds graph capture management, output-tree adaptation, shape dispatch, in-place buffer restoration, and permanent eager fallback after capture failures. |
| libreyolo/training/trainer.py | Integrates optional graph routing into both training loops while retaining eager execution for unsupported configurations and shape misses. |
| libreyolo/models/yolo9/trainer.py | Defines the captured-network/eager-loss boundary for the plain YOLO9 detection head. |
| libreyolo/models/rfdetr/trainer.py | Defines the captured RF-DETR network path while keeping target conversion and Hungarian criterion execution eager. |
| libreyolo/models/rfdetr/transformer.py | Avoids capture-time host synchronization and caches fixed spatial-shape tensors by shape and device. |
| tests/unit/test_cuda_graph_training.py | Covers dispatch and fallback behavior and verifies YOLO9 loss, parameter, and buffer parity between eager and graphed execution. |

</details>

<details open><summary><h3>Flowchart</h3></summary>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
  Batch[Training batch] --> Enabled{CUDA graph enabled and supported?}
  Enabled -- No --> Eager[Eager model forward]
  Enabled -- Yes --> Shape{Captured shape?}
  Shape -- No --> Warmup{Shape repeated enough?}
  Warmup -- No --> Eager
  Warmup -- Yes --> Snapshot[Snapshot model buffers]
  Snapshot --> Capture[Warm up and capture network]
  Capture --> Restore[Restore buffers in place]
  Restore --> Replay[Replay captured network]
  Shape -- Yes --> Replay
  Replay --> Loss[Eager family loss]
  Eager --> Loss
  Loss --> Backward[Backward and optimizer step]
```
</details>

<sub>Reviews (2): Last reviewed commit: ["Restore module buffers after capture war..."](https://github.com/libreyolo/libreyolo/commit/3ece9a044c12a15b0e95dc6a5fdd45afed1cae6f) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=48866038)</sub>

<details><summary><h4>Context used (3)</h4></summary>

- Knowledge Base — [Training Pipeline](https://app.greptile.com/libreyolo/-/custom-context/knowledge-base/libreyolo/libreyolo/-/docs/training-pipeline.md)
- Knowledge Base — [Model zoo, architectures, and inference backends](https://app.greptile.com/libreyolo/-/custom-context/knowledge-base/libreyolo/libreyolo/-/docs/model-zoo.md)
- Knowledge Base — [CLI entrypoint](https://app.greptile.com/libreyolo/-/custom-context/knowledge-base/libreyolo/libreyolo/-/docs/cli-entrypoint.md)
</details>

### CODE REVIEW

File: libreyolo/training/cuda_graph.py [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+                    sample,
+                    num_warmup_iters=self.num_warmup_iters,
+                    allow_unused_input=True,
+                )
+            except TypeError:
+                graphed = torch.cuda.make_graphed_callables(
+                    spec.network,
+                    sample,
+                    num_warmup_iters=self.num_warmup_iters,
+                )
```
[the remark this answers is not in the document]
Comment (EHxuban11) [reply libreyolo/training/cuda_graph.py:316]: Valid finding, fixed in 3ece9a04. The manager now snapshots every network buffer before capture and restores it in place after warm-up (on both success and failure paths), so BatchNorm running stats follow the exact eager trajectory; the capture batch's replay then performs that batch's single update. The YOLO9 parity test now also asserts bit-equal buffers after six optimizer steps and fails if the restore is removed.
